File: src/openaec_reports/core/tenant.py
"""Tenant configuratie — centraliseert alle asset-paden.

Multi-tenant deployment: elke klant heeft eigen templates, brand, stationery,
logo's en fonts. De engine is generiek, de look & feel per tenant configureerbaar.

Environment variables:
    OPENAEC_TENANT_DIR: Pad naar de primaire tenant directory (default brand).
    OPENAEC_TENANTS_ROOT: Pad naar parent directory met alle tenants.
        Als niet gezet, afgeleid van parent van OPENAEC_TENANT_DIR.

Fallback-chain: tenant → tenants_root/<name> → package defaults.
"""

from __future__ import annotations

import os
from pathlib import Path

# Package defaults
_PACKAGE_ASSETS = Path(__file__).parent.parent / "assets"


def detect_tenants_root() -> Path | None:
    """Detecteer de tenants root directory.

    Zoekt in volgorde:
    1. ``OPENAEC_TENANTS_ROOT`` environment variable
    2. Parent van ``OPENAEC_TENANT_DIR`` (afgeleid)

    Returns:
        Path naar tenants root, of None als niet gevonden.
    """
    env_root = os.environ.get("OPENAEC_TENANTS_ROOT")
    if env_root:
        root = Path(env_root)
        if root.exists():
            return root
    env_dir = os.environ.get("OPENAEC_TENANT_DIR")
    if env_dir:
        parent = Path(env_dir).parent
        if parent.exists() and parent != Path(env_dir):
            return parent
    return None


class TenantConfig:
    """Beheert asset-paden met fallback naar package defaults.

    Volgorde: tenant_dir → package assets.

    Usage:
        config = TenantConfig()  # leest OPENAEC_TENANT_DIR
        config = TenantConfig("/data/tenants/default")

        config.templates_dirs   # → tenant templates + package templates
        config.brand_path       # → tenant brand.yaml of package default.yaml
        config.stationery_dir   # → tenant stationery/
        config.logos_dir        # → tenant logos/
        config.fonts_dir        # → tenant fonts/
        config.tenants_root     # → parent met alle tenant directories
    """

    def __init__(self, tenant_dir: str | Path | None = None):
        env_dir = os.environ.get("OPENAEC_TENANT_DIR")
        if tenant_dir:
            self._tenant_dir = Path(tenant_dir)
        elif env_dir:
            self._tenant_dir = Path(env_dir)
        else:
            self._tenant_dir = None

    @property
    def tenant_dir(self) -> Path | None:
        return self._tenant_dir

    @property
    def tenants_root(self) -> Path | None:
        """Parent directory met alle tenant directories.

        Afgeleid van ``OPENAEC_TENANTS_ROOT`` env var, of parent van
        ``OPENAEC_TENANT_DIR``. Bruikbaar voor multi-tenant brand resolution.
        """
        env_root = os.environ.get("OPENAEC_TENANTS_ROOT")
        if env_root:
            root = Path(env_root)
            if root.exists():
                return root
        if self._tenant_dir and self._tenant_dir.parent.exists():
            parent = self._tenant_dir.parent
            if parent != self._tenant_dir:
                return parent
        return None

    @property
    def templates_dirs(self) -> list[Path]:
        """Tenant templates eerst, dan package defaults."""
        dirs = []
        if self._tenant_dir and (self._tenant_dir / "templates").exists():
            dirs.append(self._tenant_dir / "templates")
        dirs.append(_PACKAGE_ASSETS / "templates")
        return dirs

    @property
    def brand_path(self) -> Path:
        """Tenant brand.yaml, fallback naar package default.yaml."""
        if self._tenant_dir:
            tenant_brand = self._tenant_dir / "brand.yaml"
            if tenant_brand.exists():
                return tenant_brand
        return _PACKAGE_ASSETS / "brands" / "default.yaml"

    @property
    def stationery_dir(self) -> Path | None:
        if self._tenant_dir:
            d = self._tenant_dir / "stationery"
            if d.exists():
                return d
        # Fallback naar package stationery (voor backward compat)
        pkg = _PACKAGE_ASSETS / "stationery"
        if pkg.exists() and any(pkg.iterdir()):
            # Zoek eerste subdirectory met stationery bestanden
            for sub in pkg.iterdir():
                if sub.is_dir() and (sub / "standaard.pdf").exists():
                    return sub
        return None

    @property
    def logos_dir(self) -> Path | None:
        if self._tenant_dir:
            d = self._tenant_dir / "logos"
            if d.exists():
                return d
        # Fallback naar package logos (als die er zijn)
        pkg = _PACKAGE_ASSETS / "logos"
        return pkg if pkg.exists() and any(pkg.iterdir()) else None

    @property
    def fonts_dir(self) -> Path | None:
        if self._tenant_dir:
            d = self._tenant_dir / "fonts"
            if d.exists():
                return d
        pkg = _PACKAGE_ASSETS / "fonts"
        return pkg if pkg.exists() and any(pkg.iterdir()) else None

    @property
    def modules_dir(self) -> Path | None:
        """Tenant YAML module definities directory."""
        if self._tenant_dir:
            d = self._tenant_dir / "modules"
            if d.exists():
                return d
        return None

File: rust/crates/openaec-core/src/tenant.rs
//! Multi-tenant asset path management.
//!
//! Centralizes all asset paths (templates, brands, stationery, logos, fonts)
//! with a fallback chain: tenant directory -> package defaults.

use std::env;
use std::path::{Path, PathBuf};

/// Environment variable for the active tenant directory.
const ENV_TENANT_DIR: &str = "OPENAEC_TENANT_DIR";

/// Environment variable for the tenants root (parent of all tenant dirs).
const ENV_TENANTS_ROOT: &str = "OPENAEC_TENANTS_ROOT";

/// Default tenant slug used as runtime fallback when no tenant is configured.
pub const DEFAULT_TENANT: &str = "default";

/// Multi-tenant configuration with fallback to package defaults.
///
/// Asset lookup order:
/// 1. Tenant directory (if set)
/// 2. Package defaults (built-in)
#[derive(Debug, Clone)]
pub struct TenantConfig {
    /// Primary tenant directory (e.g. `tenants/default/`).
    tenant_dir: Option<PathBuf>,

    /// Root directory containing all tenant directories.
    tenants_root: Option<PathBuf>,

    /// Package assets directory (built-in defaults).
    package_assets_dir: PathBuf,
}

impl TenantConfig {
    /// Create a new TenantConfig.
    ///
    /// If `tenant_dir` is `None`, reads from `OPENAEC_TENANT_DIR` env var.
    /// Package assets are resolved relative to the crate manifest dir (for dev)
    /// or a configurable path.
    pub fn new(tenant_dir: Option<&Path>, package_assets_dir: Option<&Path>) -> Self {
        let tenant_dir = tenant_dir
            .map(PathBuf::from)
            .or_else(|| env::var(ENV_TENANT_DIR).ok().map(PathBuf::from))
            .filter(|p| p.is_dir());

        let tenants_root = detect_tenants_root(&tenant_dir);

        let package_assets_dir = package_assets_dir
            .map(PathBuf::from)
            .unwrap_or_else(default_package_assets_dir);

        Self {
            tenant_dir,
            tenants_root,
            package_assets_dir,
        }
    }

    /// The active tenant directory, if any.
    pub fn tenant_dir(&self) -> Option<&Path> {
        self.tenant_dir.as_deref()
    }

    /// The tenants root directory (parent of all tenant dirs).
    pub fn tenants_root(&self) -> Option<&Path> {
        self.tenants_root.as_deref()
    }

    /// Template directories, tenant-first then package defaults.
    ///
    /// Returns a list of directories to search for YAML templates.
    /// Templates in earlier directories take precedence.
    pub fn templates_dirs(&self) -> Vec<PathBuf> {
        let mut dirs = Vec::new();

        if let Some(ref td) = self.tenant_dir {
            let tenant_templates = td.join("templates");
            if tenant_templates.is_dir() {
                dirs.push(tenant_templates);
            }
        }

        let pkg_templates = self.package_assets_dir.join("templates");
        if pkg_templates.is_dir() {
            dirs.push(pkg_templates);
        }

        dirs
    }

    /// Path to the brand YAML file.
    ///
    /// Checks tenant `brand.yaml` first, falls back to package default.
    pub fn brand_path(&self) -> Option<PathBuf> {
        if let Some(ref td) = self.tenant_dir {
            let tenant_brand = td.join("brand.yaml");
            if tenant_brand.is_file() {
                return Some(tenant_brand);
            }
        }

        let pkg_brand = self.package_assets_dir.join("brands").join("default.yaml");
        if pkg_brand.is_file() {
            return Some(pkg_brand);
        }

        None
    }

    /// Stationery directory (PDFs/PNGs for page backgrounds).
    pub fn stationery_dir(&self) -> Option<PathBuf> {
        self.resolve_dir("stationery")
    }

    /// Logos directory.
    pub fn logos_dir(&self) -> Option<PathBuf> {
        self.resolve_dir("logos")
    }

    /// Fonts directory.
    pub fn fonts_dir(&self) -> Option<PathBuf> {
        self.resolve_dir("fonts")
    }

    /// Page types directory (tenant-only, YAML page definitions).
    pub fn page_types_dir(&self) -> Option<PathBuf> {
        self.tenant_only_dir("page_types")
    }

    /// Modules directory (tenant-only, YAML module definitions).
    pub fn modules_dir(&self) -> Option<PathBuf> {
        self.tenant_only_dir("modules")
    }

    /// Resolve a subdirectory with fallback: tenant -> package.
    fn resolve_dir(&self, name: &str) -> Option<PathBuf> {
        if let Some(ref td) = self.tenant_dir {
            let tenant_path = td.join(name);
            if tenant_path.is_dir() {
                return Some(tenant_path);
            }
        }

        let pkg_path = self.package_assets_dir.join(name);
        if pkg_path.is_dir() {
            return Some(pkg_path);
        }

        None
    }

    /// Resolve a tenant-only subdirectory (no package fallback).
    fn tenant_only_dir(&self, name: &str) -> Option<PathBuf> {
        self.tenant_dir.as_ref().and_then(|td| {
            let path = td.join(name);
            path.is_dir().then_some(path)
        })
    }
}

impl Default for TenantConfig {
    fn default() -> Self {
        Self::new(None, None)
    }
}

/// Detect the tenants root directory.
///
/// Search order:
/// 1. `OPENAEC_TENANTS_ROOT` environment variable
/// 2. Parent of the tenant directory (if set)
fn detect_tenants_root(tenant_dir: &Option<PathBuf>) -> Option<PathBuf> {
    // Check env var first.
    if let Ok(root) = env::var(ENV_TENANTS_ROOT) {
        let path = PathBuf::from(root);
        if path.is_dir() {
            return Some(path);
        }
    }

    // Derive from tenant dir (parent).
    tenant_dir
        .as_ref()
        .and_then(|td| td.parent().map(PathBuf::from))
        .filter(|p| p.is_dir())
}

/// Default package assets directory.
///
/// In development, this is `<repo>/src/openaec_reports/assets/`.
/// In a packaged context, should be configured explicitly.
fn default_package_assets_dir() -> PathBuf {
    // For now, use a relative path. In production this will be configurable.
    PathBuf::from("assets")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_config() {
        // Without any env vars or tenant dir, should not panic.
        let config = TenantConfig::new(None, None);
        assert!(config.tenant_dir().is_none());
    }

    #[test]
    fn test_with_tenant_dir() {
        let tmp = tempfile::tempdir().unwrap();
        let tenant = tmp.path().join("my_tenant");
        std::fs::create_dir_all(tenant.join("templates")).unwrap();
        std::fs::create_dir_all(tenant.join("fonts")).unwrap();
        std::fs::create_dir_all(tenant.join("stationery")).unwrap();

        let config = TenantConfig::new(Some(&tenant), None);
        assert_eq!(config.tenant_dir(), Some(tenant.as_path()));
        assert_eq!(config.templates_dirs().len(), 1);
        assert!(config.fonts_dir().is_some());
        assert!(config.stationery_dir().is_some());
    }

    #[test]
    fn test_brand_path_from_tenant() {
        let tmp = tempfile::tempdir().unwrap();
        let tenant = tmp.path().join("my_tenant");
        std::fs::create_dir_all(&tenant).unwrap();
        std::fs::write(tenant.join("brand.yaml"), "name: test").unwrap();

        let config = TenantConfig::new(Some(&tenant), None);
        let brand = config.brand_path().unwrap();
        assert!(brand.ends_with("brand.yaml"));
    }

    #[test]
    fn test_tenants_root_from_parent() {
        let tmp = tempfile::tempdir().unwrap();
        let tenants_root = tmp.path().join("tenants");
        let tenant = tenants_root.join("my_tenant");
        std::fs::create_dir_all(&tenant).unwrap();

        let config = TenantConfig::new(Some(&tenant), None);
        assert_eq!(config.tenants_root(), Some(tenants_root.as_path()));
    }

    #[test]
    fn test_tenant_only_dirs() {
        let tmp = tempfile::tempdir().unwrap();
        let tenant = tmp.path().join("tenant");
        std::fs::create_dir_all(tenant.join("modules")).unwrap();

        let config = TenantConfig::new(Some(&tenant), None);
        assert!(config.modules_dir().is_some());
        // page_types dir does not exist
        assert!(config.page_types_dir().is_none());
    }
}
